launch/doc/source/architecture.rst

```diff
@@ -0,0 +1,208 @@
+---------------------------------------
+
+The main object in `launch` is the :class:`launch.LaunchDescriptionEntity`, from which other entities which are "launched" inherit.
+This class, or more specifically classes derived from this class, are responsible for capturing the system architect's (a.k.a. the user's) intent for how the system should be launched, as well as how `launch` itself should react to asynchronous events in the system during launch.
+A launch description entity has its :meth:`launch.LaunchDescriptionEntity.visit` method called during "launching", and has any of the "describe" methods called during "introspection".
+It may also provide a :class:`asyncio.Future` with the :meth:`launch.LaunchDescriptionEntity.get_asyncio_future` method, if it has on-going asynchronous activity after returning from visit.
+
+When visited, entities may yield additional entities to be visited, and this pattern is used from the "root" of the launch, where a special entity called :class:`launch.LaunchDescription` is provided to start the launch process.
+
+The :class:`launch.LaunchDescription` class encapsulates the intent of the user as a list of discrete :class:`launch.Action`'s (which are also derived from :class:`launch.LaunchDescriptionEntity`.
+As "launch description entities" themselves, these "actions" can either be introspected for analysis without performing the sideffects, or the actions can be executed, usually in response to an event in the launch system.
+
+Additionally, launch descriptions, and the actions which they contain, can have references to :class:`launch.Substitution`'s within them.
+These substitutions are things which can be evaluated during launch and can be used to do various things like: get a launch configuration, get an environment variable, or evaluate arbitrary Python expressions.
+
+Launch descriptions, and the actions contained therein, can either be introspected directly or launched by a :class:`launch.LaunchService`.
+A launch service, is a long running activity which handles the event loop and dispatches actions.
+
+Actions
+-------
+
+The aforementioned actions allow the user to express various intentions and the set of available actions to the user can also be extended by other packages, allowing for domain specific actions.
+
+Actions can have direct side effects (e.g. run a process or set a configuration variable) and as well they can yield additional actions.
+The latter can be used to create "syntactic sugar" actions which simply yield more verbose actions.
+
+Actions may also have arguments, which can affect the behavior of the actions.
+These arguments are where :class:`launch.Substitution`'s can be used to provide more flexibility when describing reusable launch descriptions.
+
+Basic Actions
+^^^^^^^^^^^^^
+
+`launch` provides the foundational actions on which other more sophisticated actions may be built.
+This is a non-exhaustive list of actions that `launch` may provide:
```

Comment:
I also asked myself this question.

Reply:
Simply because I didn't want to have to come back here and update the list every time I added a trivial action to the list in `launch`. These are the important ones (as I see it).

Reply:
Simply because I didn't want to have to come back here and update the list every time I added a trivial action to the list in `launch`. These are the important ones (as I see it). This is just supposed to be an overview, not the complete API docs.